### PR TITLE
Implementing Triplet Loss Learning

### PR DESCRIPTION
**Pull request by dkohlsdorf**
Form the [paper](https://arxiv.org/pdf/1711.02209.pdf): 
+ Aren Jansen, Manoj Plakal, Ratheet Pandya, Daniel P. W. Ellis, :"UNSUPERVISED LEARNING OF SEMANTIC AUDIO REPRESENTATIONS"

Screen Shot 2020-09-01 at 1 01 59 AM

# Gaussian

Screen Shot 2020-09-01 at 1 02 20 AM

# Mixing

Screen Shot 2020-09-01 at 1 02 26 AM
Screen Shot 2020-09-01 at 1 02 30 AM

# In our solution

+ 1) Auto Encoder: Produces Encoder
+ 2) **Triplet Loss: Fine Tunes Encoder**
+ 3) Use encoder's pre-trained weights to train silence and type classification
+ 4) Clustering: K-means or DTW

### CODE REVIEW

**Comment by dkohlsdorf** [comment]
Model works now with the following loss:
```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to 
==================================================================================================
input_2 (InputLayer)            [(None, 128, 256, 1) 0 
__________________________________________________________________________________________________
input_3 (InputLayer)            [(None, 128, 256, 1) 0 
__________________________________________________________________________________________________
input_4 (InputLayer)            [(None, 128, 256, 1) 0 
__________________________________________________________________________________________________
model (Model)                   (None, 256)          3678208     input_2[0][0] 
                                                                 input_3[0][0] 
                                                                 input_4[0][0] 
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 768)          0           model[1][0] 
                                                                 model[2][0] 
                                                                 model[3][0] 
==================================================================================================
Total params: 3,678,208
Trainable params: 3,676,160
Non-trainable params: 2,048
__________________________________________________________________________________________________
```

and loss function:

```python
class TripletLoss(Loss):

    def __init__(self, margin, latent):
        super().__init__()
        self.margin = margin
        self.latent = latent

    def call(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        anchor = y_pred[0, 0:self.latent]
        pos    = y_pred[0, self.latent:2 * self.latent]
        neg    = y_pred[0, 2 * self.latent:3 * self.latent]
        pos_dist   = tf.reduce_sum(tf.square(tf.subtract(anchor, pos)), axis=-1)            
        neg_dist   = tf.reduce_sum(tf.square(tf.subtract(anchor, neg)), axis=-1)    
        basic_loss = tf.add(tf.subtract(pos_dist, neg_dist), self.margin)    
        loss       = tf.reduce_sum(tf.maximum(basic_loss, 0.0))               
        return loss
```

**Comment by dkohlsdorf** [comment]
Loss function ```
class TripletLoss(Loss):

    def __init__(self, margin, latent):
        super().__init__()
        self.margin = margin
        self.latent = latent

    def call(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        anchor = y_pred[:, 0:self.latent]
        pos    = y_pred[:, self.latent:2 * self.latent]
        neg    = y_pred[:, 2 * self.latent:3 * self.latent]
        pos_dist   = tf.reduce_sum(tf.square(tf.subtract(anchor, pos)), axis=-1)            
        neg_dist   = tf.reduce_sum(tf.square(tf.subtract(anchor, neg)), axis=-1)    
        basic_loss = tf.add(tf.subtract(pos_dist, neg_dist), self.margin)    
        loss       = tf.reduce_sum(tf.maximum(basic_loss, 0.0))               
        return loss
```
